File: iOS/amigoSurvey/amigoSurvey/Repository/FileUploader.swift
//
//  FileUploader.swift
//  amigoSurvey
//
//  Created by Victor Chernetsky on 10/15/17.
//  Copyright © 2017 AmigoCloud. All rights reserved.
//

import Foundation
import RxSwift
import RealmSwift

class FileUploadProgress: NSObject {
    var bytesSent: Int = 0
    var bytesTotal: Int = 0
    var message: String = ""
    var statusCode: Int = -1
    var fileIndex: Int = 0
    var filesTotal: Int = 0
    override init() {}
    
    init(bytesSent:Int, bytesTotal: Int, message: String, statusCode: Int, fileIndex: Int, filesTotal: Int) {
        self.bytesSent = bytesSent
        self.bytesTotal = bytesTotal
        self.message = message
        self.statusCode = statusCode
        self.fileIndex = fileIndex
        self.filesTotal = filesTotal
    }
}

class FileChunk: NSObject {
    var urlStr: String = ""
    var fileName: String = ""
    var data: Data?
    var ctype: String = ""
    var boundary: String = ""
    var firstByte: Int = 0
    var chunkSize: Int = 0
    var fileSize: Int = 0
    var extra: [String:String] = [:]
    var last: Bool = false
    
    init(urlStr: String,
         fileName: String,
         data: Data?,
         firstByte: Int,
         chunkSize: Int,
         fileSize: Int,
         extra: [String:String]) {
        self.urlStr = urlStr
        self.fileName = fileName
        self.data = data
        self.firstByte = firstByte
        self.chunkSize = chunkSize
        self.fileSize = fileSize
        self.extra = extra
        self.boundary = "Boundary-\(UUID().uuidString)"
        self.ctype = "multipart/form-data; boundary=\(boundary)"
    }
}

class FileUploader {
    static let sharedInstance = FileUploader()

    func getAllPhotos() -> Observable<RelatedRecord> {
        return Observable.create { observer in
            let realm = try! Realm()
            let r = realm.objects(RelatedRecord.self)
            let totalFiles = r.count
            for i in 0..<r.count {
                let record: RelatedRecord = RelatedRecord()
                record.amigo_id = r[i].value(forKey: "amigo_id")! as! String
                record.filename = r[i].value(forKey: "filename")! as! String
                record.source_amigo_id = r[i].value(forKey: "source_amigo_id")! as! String
                record.datetime = r[i].value(forKey: "datetime")! as! String
                record.location = r[i].value(forKey: "location")! as! String
                record.relatedTableId = r[i].value(forKey: "relatedTableId")! as! String
                record.recordsTotal = r[i].value(forKey: "recordsTotal")! as! Int
                record.recordsTotal = totalFiles
                observer.on(.next(record))
            }
            observer.on(.completed)
            return Disposables.create()
        }
    }
    
    func uploadAllPhotos() -> Observable<FileUploadProgress> {
        var index: Int = 0
        return FileUploader.sharedInstance.getAllPhotos()
            .flatMap{ record -> Observable<FileUploadProgress> in
                index += 1
                return self.uploadPhoto(index: index, record: record)
            }
    }
    
    func deleteAllPhotos() {
        let realm = try! Realm()
        let r = realm.objects(RelatedRecord.self)
        for i in 0..<r.count {
            let filename = r[i].value(forKey: "filename")! as! String
            deleteFile(fname: filename)
        }
        try! realm.write {
            let deletedNotifications = realm.objects(RelatedRecord.self)
            realm.delete(deletedNotifications)
        }
    }
    
    func deletePhoto(fname: String) {
        let realm = try! Realm()
        let r = realm.objects(RelatedRecord.self).filter("filename=\"\(fname)\"")
        if r != nil {
            try! realm.write {
                realm.delete(r)
            }
            deleteFile(fname: fname)
        }
    }
    
    func deleteFile(fname: String) -> Bool {
        do {
            let fileManager = FileManager.default
            let filePath = "\(SurveyConfig.getPhotoDir())/\(fname)"
            
            // Check if file exists
            if fileManager.fileExists(atPath: filePath) {
                // Delete file
                try fileManager.removeItem(atPath: filePath)
            } else {
                print("::deleteFile() File \(fname) does not exist")
            }
        }
        catch let error as NSError {
            print("An error took place: \(error)")
            return false
        }
        return true
    }
    
    func uploadPhoto(index: Int, record: RelatedRecord) -> Observable<FileUploadProgress> {
        let rt_id = CLong(record.relatedTableId)
        if let rt_id = rt_id {
            var extra: [String:String] = [:]
            extra["amigo_id"] = record.amigo_id
            extra["source_amigo_id"] = record.source_amigo_id
            //            extra["datetime"] = record.datetime
            //            extra["location"] = "\"\(record.location)\""
            
            let project = Repository.sharedInstance.findProject(projectId: SurveyConfig.getProjectId()!)
            if let project = project {
                let dataset = Repository.sharedInstance.findDataset(datasetId: SurveyConfig.getDatasetId()!, project: project)
                if let dataset = dataset {
                    let rt = Repository.sharedInstance.findRelatedTable(related_table_id: rt_id, dataset: dataset)
                    if let rt = rt {
                        return FileUploader.sharedInstance.chunkedFileUpload(url: rt.chunked_upload,
                                                                              url_complete: rt.chunked_upload_complete,
                                                                              path: SurveyConfig.getPhotoDir(),
                                                                              fname: record.filename,
                                                                              relatedTableId: record.relatedTableId,
                                                                              extra: extra,
                                                                              chunkSize: 500000,
                                                                              fileIndex: index,
                                                                              filesTotal: record.recordsTotal)
                    }
                }
            }
        }
        return Observable<FileUploadProgress>.just(FileUploadProgress())
    }

    func createBodyForFileUpload(parameters: [String: String],
                                 boundary: String,
                                 data: Data?,
                                 mimeType: String,
                                 filename: String) -> Data {
        let body = NSMutableData()
        
        let boundaryPrefix = "--\(boundary)\r\n"
        
        for (key, value) in parameters {
            body.appendString(boundaryPrefix)
            body.appendString("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n")
            body.appendString("\(value)\r\n")
        }
        
        body.appendString(boundaryPrefix)
        body.appendString("Content-Disposition: form-data; name=\"datafile\"; filename=\"\(filename)\"\r\n")
        body.appendString("Content-Type: \(mimeType)\r\n\r\n")
        if data != nil {
            body.append(data!)
        }
        body.appendString("\r\n")
        body.appendString("--".appending(boundary.appending("--")))
        
        return body as Data
    }
    
    func paramsToUrlEncoded(params: [String:String]) -> String {
        var str = ""
        var count = 0
        for (key, value) in params {
            if count > 0 {
                str += "&"
            }
            str += "\(key)=\(value)"
            count += 1
        }
        return str
    }
    
    func chunkedFileUpload(url: String,
                           url_complete:String,
                           path: String,
                           fname: String,
                           relatedTableId: String,
                           extra: [String:String],
                           chunkSize: Int,
                           fileIndex: Int,
                           filesTotal: Int) -> Observable<FileUploadProgress> {
        var upload_id = ""
        var statusCode = -1
        return self.getFileChunks(url: url, path: path, fname: fname, chunkSize: chunkSize, extra: extra)
            .flatMap { chunk -> Observable<FileUploadProgress> in
                    if !chunk.last {
                        chunk.extra["upload_id"] = upload_id
                        upload_id = self.postFileChunk(chunk: chunk)
                        if upload_id == "" {
                            statusCode = -1
                        }
                    } else {
                        var params = extra
                        params["md5"] = "90affbd9a1954ec9ff029b7ad7183a16" // Bogus value
                        params["filename"] = fname
                        params["upload_id"] = upload_id
                        let data = self.paramsToUrlEncoded(params: params)
                        let escdata = data.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
                        chunk.data = escdata.data(using: String.Encoding.utf8)
                        chunk.ctype = "application/x-www-form-urlencoded"
                        chunk.urlStr = url_complete
                        statusCode = self.postFileComplete(chunk: chunk)
                    }
                    return Observable.from(FileUploadProgress(bytesSent: chunk.firstByte,
                                                              bytesTotal: chunk.fileSize,
                                                              message: chunk.fileName,
                                                              statusCode: statusCode,
                                                              fileIndex: fileIndex,
                                                              filesTotal: filesTotal))
                }
    }
    
    func getFileChunks(url: String, path: String, fname: String, chunkSize: Int, extra: [String:String]) -> Observable<FileChunk> {
        var lastChunk: FileChunk? = nil
        return Observable.create { observer in
            let fullPath = "\(path)/\(fname)"
            if let data = NSData(contentsOfFile: fullPath) {
                let fileSize = data.length
                var bytesRead: Int = 0
                var buffer: [UInt8] = Array(repeating: 0, count: fileSize)
                data.getBytes(&buffer, range: NSRange(location: 0, length: fileSize))
                repeat {
                    let bytesLeft = fileSize - bytesRead
                    var thisChunkSize = chunkSize
                    if bytesLeft < chunkSize {
                        thisChunkSize = bytesLeft
                    }
                    lastChunk = FileChunk(urlStr:url,
                                          fileName: fname,
                                          data: Data(buffer[bytesRead..<bytesRead+thisChunkSize]),
                                          firstByte: bytesRead,
                                          chunkSize: thisChunkSize,
                                          fileSize: fileSize,
                                          extra: extra)
                    if let lastChunk = lastChunk {
                        observer.on(.next(lastChunk))
                    }
                    bytesRead += thisChunkSize
                } while bytesRead < fileSize
                if let lastChunk = lastChunk {
                    lastChunk.last = true
                    observer.on(.next(lastChunk))
                }
                observer.on(.completed)
            }
            return Disposables.create()
        }
    }

    func postFileChunk(chunk: FileChunk) -> String  {
        var upload_id = ""
        let url = URL(string: chunk.urlStr)
        let config = URLSessionConfiguration.default
        let request = NSMutableURLRequest(url: url!)
        request.httpMethod = "POST"
        request.setValue("Bearer " + AmigoRest.sharedInstance.amigoToken.access_token, forHTTPHeaderField: "Authorization")
        request.setValue(chunk.ctype, forHTTPHeaderField: "Content-Type")
        let content_range = "bytes \(chunk.firstByte)-\(chunk.firstByte+chunk.chunkSize-1)/\(chunk.fileSize)"
        request.setValue(content_range, forHTTPHeaderField: "Content-Range")
        request.httpBody = self.createBodyForFileUpload(parameters: chunk.extra,
                                                        boundary: chunk.boundary,
                                                        data: chunk.data,
                                                        mimeType: "image/*",
                                                        filename: chunk.fileName)
        let semaphore = DispatchSemaphore(value: 0)
        let session = URLSession(configuration: config, delegate: AmigoRest.sharedInstance, delegateQueue: OperationQueue.current)
        let task = session.dataTask(with: request as URLRequest, completionHandler: {(data, response, error) in
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            if statusCode == 200 && data != nil {
                let json = try! JSONSerialization.jsonObject(with: data as! Data, options: []) as! [String: Any]
                upload_id = json["upload_id"] as! String
            } else {
//                print("Response: \(response.debugDescription)")
            }
            semaphore.signal()
            } as (Data?, URLResponse?, Error?) -> Void)
        task.resume()
        semaphore.wait(timeout: DispatchTime.distantFuture)
        return upload_id
    }
    
    func postFileComplete(chunk: FileChunk) -> Int {
        let url = URL(string: chunk.urlStr)
        let config = URLSessionConfiguration.default
        let request = NSMutableURLRequest(url: url!)
        request.httpMethod = "POST"
        request.setValue("Bearer " + AmigoRest.sharedInstance.amigoToken.access_token, forHTTPHeaderField: "Authorization")
        request.setValue(chunk.ctype, forHTTPHeaderField: "Content-Type")
        request.httpBody = chunk.data
        var statusCode = -1
        let semaphore = DispatchSemaphore(value: 0)
        let session = URLSession(configuration: config, delegate: AmigoRest.sharedInstance, delegateQueue: OperationQueue.current)
        let task = session.dataTask(with: request as URLRequest, completionHandler: {(data, response, error) in
            statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            semaphore.signal()
            } as (Data?, URLResponse?, Error?) -> Void)
        task.resume()
        semaphore.wait(timeout: DispatchTime.distantFuture)
        return statusCode
    }
 
}

File: Android/AmigoSurvey/app/src/main/java/com/amigocloud/amigosurvey/form/FileUploader.kt
package com.amigocloud.amigosurvey.form

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.LiveDataReactiveStreams
import android.arch.lifecycle.ViewModel
import com.amigocloud.amigosurvey.models.RelatedRecord
import com.amigocloud.amigosurvey.repository.AmigoRest
import com.amigocloud.amigosurvey.repository.Repository
import com.amigocloud.amigosurvey.repository.SurveyConfig
import com.amigocloud.amigosurvey.viewmodel.INFLATION_EXCEPTION
import com.amigocloud.amigosurvey.viewmodel.ViewModelFactory
import com.squareup.moshi.Moshi
import io.reactivex.BackpressureStrategy
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.processors.BehaviorProcessor
import okhttp3.MediaType
import okhttp3.RequestBody
import java.io.File
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

data class ChunkedUploadResponse (var upload_id:String = "",
                                  var offset: Int = 0,
                                  var expires: String = "")

sealed class FileProgressEvent(val message: String,
                               val fileIndex: Int,
                               val filesTotal: Int,
                               val record: RelatedRecord?)

class FileUploadProgressEvent(
        message: String = "",
        fileIndex: Int = 0,
        filesTotal: Int = 0,
        record: RelatedRecord? = null,
        val bytesSent: Int = 0,
        val bytesTotal: Long = 0
) : FileProgressEvent(message, fileIndex, filesTotal, record)

class FileUploadCompleteEvent(message: String = "",
                              fileIndex: Int = 0,
                              filesTotal: Int = 0,
                              record: RelatedRecord? = null) : FileProgressEvent(message, fileIndex, filesTotal, record)

data class FileChunk (
        var fileName: String = "",
        var urlStr: String = "",
        var data: ByteArray? = null,
        var firstByte: Int = 0,
        var chunkSize: Int = 0,
        var fileSize: Long = 0,
        var record: RelatedRecord? = null,
        var last: Boolean = false,
        var boundary: String = "Boundary-${UUID.randomUUID()}",
        var ctype: String = "multipart/form-data; boundary=${boundary}"
)

class FileUploader(private val rest: AmigoRest,
                   private val repository: Repository,
                   private val config: SurveyConfig): ViewModel() {

    private val processor = BehaviorProcessor.create<Pair<Long, Long>>()

    val events: LiveData<FileProgressEvent> = LiveDataReactiveStreams.fromPublisher(processor
            .flatMap { (pId, dId) -> uploadPhotos(pId, dId).toFlowable(BackpressureStrategy.LATEST) })

    fun uploadAllPhotos(projectId: Long, datasetId: Long) {
        processor.onNext(projectId.to(datasetId))
    }

    private fun uploadPhotos(projectId: Long, datasetId: Long): Observable<FileProgressEvent> {
        var index = 0
        val totalPhotos = repository.relatedRecordDao().all.size
        return if (totalPhotos == 0) {
            Observable.just(FileUploadCompleteEvent("No files to upload", -1, 0, null))
        } else {
            getAllPhotos().concatMap { uploadPhoto(index++, it, projectId, datasetId, totalPhotos) }
        }
    }

    fun deletePhoto(record: RelatedRecord) {
        repository.relatedRecordDao().delete(record)
    }

    private fun getAllPhotos(): Observable<RelatedRecord> {
        return Observable.create {
            for (record in repository.relatedRecordDao().all) {
                it.onNext(record)
            }
        }
    }

    fun getSavedPhotosNum(): Int {
        val records = repository.relatedRecordDao().all
        return records.count()
    }

    private fun uploadPhoto(index: Int, record: RelatedRecord, projectId: Long, datasetId: Long, totalPhotos: Int): Observable<FileProgressEvent> {
        return rest.fetchRelatedTables(projectId, datasetId)
                .flatMapObservable { Observable.fromIterable(it.results) }
                .filter { it.id == record.relatedTableId.toLong() }
                .flatMap {
                    chunkedFileUpload(it.chunked_upload,
                            it.chunked_upload_complete,
                            config.photosDir,
                            record,
                            500000,
                            index,
                            totalPhotos)
                }
    }

    private fun chunkedFileUpload(url: String,
                          url_complete:String,
                          path: String,
                          record: RelatedRecord,
                          chunkSize: Int,
                          fileIndex: Int,
                          filesTotal: Int) : Observable<FileProgressEvent> {
        val md5 = "90affbd9a1954ec9ff029b7ad7183a16" // Bogus value
        return getFileChunks(url, path, record.filename, chunkSize, record)
                .filter { it.isNotEmpty() }
                .flatMapSingle { chunks ->
                    postFileChunkFirst(chunks[0]).map { chunks.to(chunks[0].to(it)) }
                }
                .flatMapObservable { (chunks, firstChunk) ->
                    val uploadId = firstChunk.second.upload_id
                    val observables = mutableListOf<Observable<FileChunk>>(Observable.just(firstChunk.first))
                    var lastChunk : FileChunk? = null
                    chunks.forEachIndexed { index, chunk ->
                        if (chunk.last) lastChunk = chunk
                        if(index > 0) {
                            observables.add(postFileChunk(uploadId, chunk)
                                    .toObservable().map { chunk })
                        }
                    }
                    lastChunk?.let { lc ->
                        lc.urlStr = url_complete
                        lc.ctype = "application/x-www-form-urlencoded"
                        observables.add(postFileComplete(uploadId, md5, lc)
                                .toObservable().map { lc })

                        Observable.concat(observables)
                    }
                }
                .map { fileChunk ->
                    if ( (fileChunk.firstByte + fileChunk.chunkSize).toLong() == fileChunk.fileSize) {
                        FileUploadCompleteEvent("fileIndex:${fileIndex}, filesTotal:${filesTotal}, uploaded bytes:${fileChunk.firstByte + fileChunk.chunkSize}, fileSize:${fileChunk.fileSize}",
                                fileIndex,
                                filesTotal,
                                fileChunk.record)
                    } else {
                        FileUploadProgressEvent(
                                record.filename,
                                fileIndex,
                                filesTotal,
                                fileChunk.record,
                                fileChunk.firstByte + fileChunk.chunkSize,
                                fileChunk.fileSize)
                    }
                }
    }

    private fun getFileChunks(url: String, path: String, fname: String, chunkSize: Int, record: RelatedRecord): Single<List<FileChunk>> {
        return Single.create { emitter ->
            val file = File("$path/$fname")
            val totalBytes = file.length()
            var bytesRead = 0
            val chunkList = mutableListOf<FileChunk>()
            file.forEachBlock(blockSize = chunkSize, action = { buffer, count ->
                        bytesRead += count
                        val isLast = (bytesRead >= totalBytes)
                        chunkList.add(FileChunk(fname, url, buffer.clone(), bytesRead-count, count, totalBytes, record, isLast))
                    })
            emitter.onSuccess(chunkList)
        }
    }

    private fun postFileChunkFirst(chunk: FileChunk): Single<ChunkedUploadResponse> {
        val body = RequestBody.create(MediaType.parse("application/octet-stream"), chunk.data, 0, chunk.chunkSize)

        val bodyMap = mutableMapOf(
                Pair("amigo_id", RequestBody.create(MediaType.parse("multipart/form-data"), chunk.record?.amigo_id!!)),
                Pair("source_amigo_id", RequestBody.create(MediaType.parse("multipart/form-data"), chunk.record?.source_amigo_id!!)),
                Pair("filename", RequestBody.create(MediaType.parse("multipart/form-data"), chunk.fileName))
                ,Pair("datafile\"; filename=\"${chunk.fileName}", body)
        )

        val headers = mapOf(
                Pair("Content-Range", "bytes ${chunk.firstByte}-${chunk.firstByte+chunk.chunkSize-1}/${chunk.fileSize}"))
        return rest.chunkedUploadFirst(
                chunk.urlStr,
                bodyMap,
                headers)
    }

    private fun postFileChunk(upload_id: String, chunk: FileChunk): Single<ChunkedUploadResponse> {
        val body = RequestBody.create(MediaType.parse("application/octet-stream"), chunk.data, 0, chunk.chunkSize)

        val bodyMap = mutableMapOf(
                Pair("amigo_id", RequestBody.create(MediaType.parse("multipart/form-data"), chunk.record?.amigo_id!!)),
                Pair("source_amigo_id", RequestBody.create(MediaType.parse("multipart/form-data"), chunk.record?.source_amigo_id!!)),
                Pair("filename", RequestBody.create(MediaType.parse("multipart/form-data"), chunk.fileName)),
                Pair("upload_id", RequestBody.create(MediaType.parse("multipart/form-data"), upload_id))
                ,Pair("datafile\"; filename=\"${chunk.fileName}", body)
        )

        val headers = mapOf(
                Pair("Content-Range", "bytes ${chunk.firstByte}-${chunk.firstByte+chunk.chunkSize-1}/${chunk.fileSize}"))
        return rest.chunkedUpload(
                chunk.urlStr,
                bodyMap,
                headers)
    }

    private fun postFileComplete(upload_id: String, md5: String, chunk: FileChunk): Single<ChunkedUploadResponse> {
        chunk.record?.let { record ->
            return rest.chunkedUploadComplete(
                    chunk.urlStr,
                    upload_id,
                    md5,
                    record.source_amigo_id,
                    chunk.fileName,
                    mapOf(Pair("Content-Type", chunk.ctype)))
        }
        return Single.just(ChunkedUploadResponse())
    }

    @Suppress("UNCHECKED_CAST")
    class Factory @Inject constructor(private val rest: AmigoRest,
                                      private val repository: Repository,
                                      private val config: SurveyConfig) : ViewModelFactory<FileUploader>() {

        override val modelClass = FileUploader::class.java

        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            if(modelClass.isAssignableFrom(this.modelClass)) {
                return FileUploader(rest, repository, config) as T
            }
            throw IllegalArgumentException(INFLATION_EXCEPTION)
        }
    }
}
